starlark/profile.go

```diff
@@ -0,0 +1,419 @@
+// Copyright 2019 The Bazel Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package starlark
+
+// This file defines a simple execution-time profiler for Starlark.
+// It measures the wall time spent executing Starlark code, and emits a
+// gzipped protocol message in pprof format (github.com/google/pprof).
+//
+// When profiling is enabled, the interpreter calls the profiler to
+// indicate the start and end of each "span" or time interval. A leaf
+// function (whether Go or Starlark) has a single span. A function that
+// calls another function has spans for each interval in which it is the
+// top of the stack. (A LOAD instruction also ends a span.)
+//
+// At the start of a span, the interpreter records the current time in
```

Comment:
It sounds like this will measure time spent inside Go code, but it can't record stacks inside Go.

Do you have a sense of what the overhead is for having the profiler on? Calling `gettimeofday` on every span seems like it would be expensive. How many times does that typically happen inside a quantum?

Since profiling is being driven by the interpreter, maybe there are other options available? The interpreter could count instructions executed and check the time after ~1000 instructions (and before/after every call to Go code). That could be checked on back-branch and call instructions to avoid branch in the main interpreter loop.

Reply:
If the profiler overhead is less than 10%, then this is probably overengineering, so feel free to disregard.

The counting instructions idea would reduce the number of calls to gettimeofday. You don't actually need to count instructions executed, since you're still measuring time. Just increment a counter when you enter a function or take a loop back branch. When the counter reaches some threshold, that's when you call gettimeofday.

Reply:
The problem is that any function call could potentially take a long time, so we need to inspect the clock at least at every call return (and after a module load instruction) otherwise we're likely to sample the stack only after the most expensive frame has been popped. The vdso implementation of gettime is fast enough (19ns) that even redundantly calling it twice per span is not that slow. (Redundant because since a function call entails end-start in the caller and start-end in the callee, which is four clock ops even though two would suffice.)